src/server/acl/acl_family.cc

```diff
@@ -57,10 +61,113 @@ void AclFamily::List(CmdArgList args, ConnectionContext* cntx) {
   }
 }
 
+namespace {
+
+std::optional<std::string> MaybeParsePassword(std::string_view command) {
+  if (command[0] != '>') {
+    return {};
+  }
+
+  return {std::string(command.substr(1))};
+}
+
+std::optional<bool> MaybeParseStatus(std::string_view command) {
+  if (command == "ON") {
+    return true;
+  }
+  if (command == "OFF") {
+    return false;
+  }
+  return {};
+}
+
+using OptCat = std::optional<uint32_t>;
+
+std::pair<OptCat, OptCat> MaybeParseAclCategory(std::string_view command) {
+  if (command[0] == '+' && command[1] == '@') {
+    auto res = CATEGORY_INDEX_TABLE.find(command.substr(2));
+    if (res == CATEGORY_INDEX_TABLE.end()) {
+      return {};
+    }
+    return {res->second, {}};
+  }
+
+  if (command[0] == '-' && command[1] == '@') {
+    auto res = CATEGORY_INDEX_TABLE.find(command.substr(2));
+    if (res == CATEGORY_INDEX_TABLE.end()) {
+      return {};
+    }
+    return {{}, res->second};
+  }
+
+  return {};
+}
+
+std::variant<User::UpdateRequest, std::string> ParseAclSetUser(CmdArgList args) {
+  User::UpdateRequest req;
+
+  auto maybe_update_acl = [](auto& output, const auto& input) {
+    if (input) {
+      if (!output) {
+        output = *input;
+      }
+      *output |= *input;
+    }
+  };
+
+  for (auto arg : args) {
+    ToUpper(&arg);
+    const auto command = facade::ToSV(arg);
+    if (auto pass = MaybeParsePassword(command); pass) {
+      if (req.password) {
+        return "Only one password is allowed";
+      }
+      req.password = std::move(pass);
+      continue;
+    }
+
+    if (auto status = MaybeParseStatus(command); status) {
+      if (req.is_active) {
+        return "Multiple ON/OFF are not allowed";
+      }
+      req.is_active = *status;
+      continue;
+    }
+
+    auto [acl_plus, acl_minus] = MaybeParseAclCategory(command);
+    if (!acl_plus && !acl_minus) {
+      using namespace std::string_literals;
+      return "Unrecognized parameter: "s + std::string(facade::ToSV(arg));
+    }
+
+    maybe_update_acl(req.plus_acl_categories, acl_plus);
+    maybe_update_acl(req.minus_acl_categories, acl_minus);
```

Comment:
🤯 🔫 

So if you choose to use the optional<[cat, bool remove]> you can just
```c++
if (result) {
  auto* acl_field = result.second ? &req.minus_acl_categories : req.plus_acl_categories;
  *acl_field |=  result.first;
} else {
  return ErrorReply{absl::StrCat("Unrecognized paramter", command)};
}
```

Reply:
If acl_field is an optional you can do `*acl_field = acl_field -> value_or(0) | result.first ` instead of adding if's 🤓

Reply:
This is too good :gem: :gem: